### ISSUE TITLE
File Uploads Not Working

### ISSUE THREAD

**Issue by JordanDalton** [Sammyjo20/Saloon#324]
I'm using version 2.

Scenario: I need to POST to an API endpoint the following values:

- `reference_id` : String
- `files`: Array of Files

Here is my request file, with redactions:

```
<?php

namespace {redacted}\{redacted}\Requests;

use Saloon\Contracts\Body\HasBody;
use Saloon\Data\MultipartValue;
use Saloon\Enums\Method;
use Saloon\Http\Request;
use Saloon\Traits\Body\HasJsonBody;
use Saloon\Traits\Body\HasMultipartBody;

class CreateClientRequest extends Request implements HasBody
{
    use HasMultipartBody;

    protected Method $method = Method::POST;

    public function __construct(public string|int $reference_id, public array $files = [])
    {
        //
    }

    public function resolveEndpoint(): string
    {
        return '/clients';
    }

    protected function defaultHeaders(): array
    {
        $boundary = md5(uniqid());

        return [
            'Content-Type' => "multipart/form-data; boundary=$boundary; charset=utf-8"
        ];
    }

    protected function buildMultipartValues(): array
    {
        $multipart_values = [];

        foreach($this->files as $file)
        {
            $multipart_values[] = new MultipartValue(
                    name: $file['name'],
                   value: file_get_contents($file['path']),
                filename: $file['filename'],
                 headers: [
                    'Content-Type' => 'text/plain'
                ]
            );
        }

        return $multipart_values;
    }

    protected function defaultBody(): array
    {
        $default = [new MultipartValue(name: 'reference_id', value: $this->reference_id)];
        $files   = $this->buildMultipartValues();

        return $default + $files;
    }
}

```

---

I'm following the documentation at: https://docs.saloon.dev/v/2/the-basics/request-body-data/multipart-form-body

The documentation states:

`Saloon will automatically send the Content-Type: multipart/form-data header for you when using the HasMultipartBody trait, however, if you would like to overwrite this behaviour then you can use the defaultHeaders method on the request or modify the headers before the request is sent.`

However, the receiving API does not show a multipart/form-data header which is why you'll see my use of `defaultHeaders()`.

The receiving API, a Laravel application, is validating the request with the following rules:

```
$request->validate([
          'reference_id' => [
              'required',
              'string',
              'max:255',
              Rule::unique('api_clients', 'reference_id')
                  ->where(
                      'api_organization_id',
                      $request->organization()->id
                  )
          ],
          'files'   => 'required|array',
          'files.*' => 'file|mimes:txt',
     ]);
```

Upon sending the request I receive:

```
  "message" => "The reference id field is required. (and 1 more error)"
  "errors" => array:2 [
    "reference_id" => array:1 [
      0 => "The reference id field is required."
    ]
    "files" => array:1 [
      0 => "The files field is required."
    ]
  ]
]
```

I've tried variations of the request (using different traits, defaultQuery(), etc).

Not sure if this helps but this is the guzzle sample Postman provided me when I successfully issued my API request with Postman:

```
<?php
$client = new Client();
$headers = [
  '{redacted}-Api-Key' => '{redacted}',
  'Accept' => 'application/json'
];
$options = [
  'multipart' => [
    [
      'name' => 'reference_id',
      'contents' => '123-456-789'
    ],
    [
      'name' => 'files[0]',
      'contents' => Utils::tryFopen('/Users/jordandalton/Code/Commercial/{redacted}/packages/package-skeleton-php/tests/tmp/test-file.txt', 'r'),
      'filename' => '/Users/jordandalton/Code/Commercial/{redacted}/packages/package-skeleton-php/tests/tmp/test-file.txt',
      'headers'  => [
        'Content-Type' => '<Content-type header>'
      ]
    ]
]];
$request = new Request('POST', 'https://{redacted}.test/api/v1/clients', $headers);
$res = $client->sendAsync($request, $options)->wait();
echo $res->getBody();
```

**Comment by Sammyjo20** [comment]
Hey @JordanDalton! Thanks for opening an issue.

Looking at your code I see one thing which might be causing problems - are you able to try removing your custom `Content-Type` and see if it works? You shouldn't need to set your own boundary as Guzzle will do that automatically for you behind the scenes, all you need to do is build up the `MultipartValue` classes. I think what's happening is that Guzzle is internally creating a boundary of its own for each of the values which doesn't correspond with the boundary that you are overwriting, which is in turn causing your receiving application to not see any files. 

This is something which is improved upon in v3 because Saloon builds thee multipart stream, not Guzzle.

**Comment by JordanDalton** [comment]
Thank you for looking into this.

So I made the changes based on your recommendation:

Here's my updated code:

```php
<?php

namespace BiglySales\BiglySalesAiSdk\Requests;

use Saloon\Contracts\Body\HasBody;
use Saloon\Data\MultipartValue;
use Saloon\Enums\Method;
use Saloon\Http\Request;
use Saloon\Traits\Body\HasJsonBody;
use Saloon\Traits\Body\HasMultipartBody;

class CreateClientRequest extends Request implements HasBody
{
    use HasMultipartBody;

    protected Method $method = Method::POST;

    public function __construct(public string|int $reference_id, public array $files = [])
    {
        //
    }

    public function resolveEndpoint(): string
    {
        return '/clients';
    }

    protected function defaultBody(): array
    {
        $bodies = [];

        // Add the reference id.
        $bodies[] = new MultipartValue(
             name: 'reference_id',
            value: $this->reference_id
        );

        // Add the various files.
        foreach ($this->files as $k => $file)
        {
            $bodies[] = new MultipartValue(
                    name: "files[$k]",
                   value: file_get_contents($file['path']),
                filename: $file['filename'],
                 headers: [
                    'Content-Type' => 'text/plain',
                ]
            );
        }

        return $bodies;
    }
}

```




however, the receiving API still reports that the reference_id and files are missing.

```phparray:2 [
  "message" => "The reference id field is required. (and 1 more error)"
  "errors" => array:2 [
    "reference_id" => array:1 [
      0 => "The reference id field is required."
    ]
    "files" => array:1 [
      0 => "The files field is required."
    ]
  ]
]
```

I did confirm the MultipartBodyRepository as data but it seems that Laravel doesn't recognize them coming across:

```
Saloon\Repositories\Body\MultipartBodyRepository {#726
  #data: Saloon\Repositories\Body\ArrayBodyRepository {#723
    #data: array:2 [
      "reference_id" => Saloon\Data\MultipartValue {#725
        +name: "reference_id"
        +value: "test-654c2375e25dd"
        +filename: null
        +headers: []
      }
      "files[0]" => Saloon\Data\MultipartValue {#724
        +name: "files[0]"
        +value: "Hello World"
        +filename: "test-file.txt"
        +headers: array:1 [
          "Content-Type" => "text/plain"
        ]
      }
    ]
  }
}
```

**Comment by JordanDalton** [comment]
Here's the dump from the middleware:

$connector = ->sender()->addMiddleware(function (callable $handler) {
        return function ($request, array $options) use ($handler) {
            dd($request->getBody());
        };
    });

```
GuzzleHttp\Psr7\MultipartStream {#684
  -boundary: "95b3a8c5fde0c42900fe2a972dee70193f79f631"
  -stream: GuzzleHttp\Psr7\AppendStream {#683
    -streams: array:4 [
      0 => GuzzleHttp\Psr7\Stream {#681
        -stream: stream resource {@645
          wrapper_type: "PHP"
          stream_type: "TEMP"
          mode: "w+b"
          unread_bytes: 0
          seekable: true
          uri: "php://temp"
          options: []
        }
        -size: 121
        -seekable: true
        -readable: true
        -writable: true
        -uri: "php://temp"
        -customMetadata: []
      }
      1 => GuzzleHttp\Psr7\Stream {#682
        -stream: stream resource {@642
          wrapper_type: "PHP"
          stream_type: "TEMP"
          mode: "w+b"
          unread_bytes: 0
          seekable: true
          uri: "php://temp"
          options: []
        }
        -size: 4
        -seekable: true
        -readable: true
        -writable: true
        -uri: "php://temp"
        -customMetadata: []
      }
      2 => GuzzleHttp\Psr7\Stream {#680
        -stream: stream resource {@647
          wrapper_type: "PHP"
          stream_type: "TEMP"
          mode: "w+b"
          unread_bytes: 0
          seekable: true
          uri: "php://temp"
          options: []
        }
        -size: 2
        -seekable: true
        -readable: true
        -writable: true
        -uri: "php://temp"
        -customMetadata: []
      }
      3 => GuzzleHttp\Psr7\Stream {#679
        -stream: stream resource {@649
          wrapper_type: "PHP"
          stream_type: "TEMP"
          mode: "w+b"
          unread_bytes: 0
          seekable: true
          uri: "php://temp"
          options: []
        }
        -size: 46
        -seekable: true
        -readable: true
        -writable: true
        -uri: "php://temp"
        -customMetadata: []
      }
    ]
    -seekable: true
    -current: 0
    -pos: 0
  }
}

```


```
--6b08fb375441a12720464c968df124ff82ee9945\r\n
X-Baz: bar\r\n
Content-Disposition: form-data; name="foo"\r\n
Content-Length: 4\r\n
\r\n
data\r\n
--6b08fb375441a12720464c968df124ff82ee9945--\r\n
```

Receiving API shows no data.

**Comment by JordanDalton** [comment]
I've been tinkering with it, but on Laravel's end `$request->all()` show:

```
array:2 [
  0 => array:4 [
    "name" => "reference_id"
    "contents" => "test-654ccf60a87b8"
    "filename" => null
    "headers" => []
  ]
  1 => array:4 [
    "name" => "files[0]"
    "contents" => "Hello World"
    "filename" => "test-file.txt"
    "headers" => array:1 [
      "Content-Type" => "text/plain"
    ]
  ]
]
```

As you'll see Laravel receives it as an array which $request->get('reference_id') won't work

**Comment by OthmanAdlan** [comment]
@Sammyjo20 @JordanDalton 

I encountered the same issue when attempting to send Multipart form and make it work in v3 without success. However, after downgrading to v2, it is now operating smoothly.

**Comment by Sammyjo20** [comment]
Hey @OthmanAdlan did you also have a mixture of files/content? I've been using Multipart uploads in both v2+v3 in my Laravel projects and it's been fine for uploading a mix. Maybe I'm missing something in the docs.

**Comment by JordanDalton** [comment]
I got it to work, but here's how:

1) Make sure connector class default headers doesn't override the Content-Type;

Here's my request class:

```php
<?php

namespace {redacted}\{redacted}\Requests;

use Saloon\Data\MultipartValue;
use Saloon\Enums\Method;
use Saloon\Http\Request;

class CreateClientRequest extends Request
{
    protected Method $method = Method::POST;

    public function __construct(public string|int $reference_id, public array $files = [])
    {
        //
    }

    public function resolveEndpoint(): string
    {
        return '/clients';
    }

    public function defaultConfig(): array
    {
        $multipart = [];

        foreach ($this->files as $k => $file)
        {
            $multipart[] = (new MultipartValue(
                    name: "files[$k]",
                   value: file_get_contents($file['path']),
                filename: $file['filename'],
                 headers: [
                    'Content-Type' => 'text/plain',
                ]
            ))->toArray();
        }

        $multipart[] = [
            'name'     => 'reference_id',
            'contents' => $this->reference_id
        ];

        return [
            'multipart' => $multipart
        ];
    }
}
```

**Comment by OthmanAdlan** [comment]
@Sammyjo20
Yeah actually my request does have a mixture of a file and another string field.
```
protected function defaultBody(): array
{
    return [
        new MultipartValue(name: 'type', value: 'user_avatar'),
        new MultipartValue(name: 'file', value: fopen(public_path('img/avatar.png'), 'r')),
    ];
}
```

**Comment by Sammyjo20** [comment]
Thanks for letting me know @JordanDalton and @OthmanAdlan - I wonder if it's something to do with mixed values! I will setup a test bench and see if I can replicate it

**Comment by Sammyjo20** [comment]
I have been struggling to replicate this in v3. The only way I could replicate it was if I didn't provide a filename, the receiving Laravel app would say 'The file must be a file'. @JordanDalton are you able to send me the equivalent Laravel rules so I can replicate it on my machine? Also @OthmanAdlan if there's anything more you can share like if you are also using consuming Laravel API that would be useful to know.

**Comment by OthmanAdlan** [comment]
@Sammyjo20 

I am utilizing Laravel on both ends - Saloon for sending the request and the API for receiving the request of the uploaded file.

These are the rules I am implementing on the API that handles the incoming request:
```
    'file' => ['required', 'image', 'mimes:png,jpg,jpeg', 'max:2048'],
    'type' => ['required', 'string'],
```

**Comment by JordanDalton** [comment]
I'd need to pull in the v3 branch to test. I was developing against v2.

**Comment by Sammyjo20** [comment]
Thanks both - seems like it's okay on v3. Let me know if you can confirm that Jordan 🤠 I will test against v2 later this week and hopefully find the issue!

**Comment by ManukMinasyan** [comment]
https://flareapp.io/share/qm1e8Dwm - @Sammyjo20 How I can use multiple values for HasMultipartBody?

**Comment by Sammyjo20** [comment]
@ManukMinasyan Please can you open a separate issue.

**Comment by Sammyjo20** [comment]
Sorry for being slow on this one Jordan. I'm fairly certain this is an edge case, but I'm still trying to replicate it.

**Comment by JordanDalton** [comment]
You're good brother. Things on my end are resolved :)

**Comment by Sammyjo20** [comment]
I'm going to close this issue for now as it's for v2 and I've just released #340 for v3 which should improve compatibility when needing to send multiple multipart values with the same name. Feel free to re-open.
